mitodl/open-discussions
# Enforce uniqueness for instructor full name, get rid of existing dupes

#### Pre-Flight checklist



#### What are the relevant tickets?
Closes #3716

#### What's this PR do?
Enforces uniqueness on the `CourseInstructor.full_name` field, removes existing dupes and updates the list of instructors for any affected runs.

#### How should this be manually tested?
Add the following to your `.env` file:
```
MITX_ONLINE_BASE_URL=https://mitxonline.mit.edu/
MITX_ONLINE_COURSES_API_URL=https://mitxonline.mit.edu/api/courses/
MITX_ONLINE_PROGRAMS_API_URL=https://mitxonline.mit.edu/api/programs/
MITX_ONLINE_LEARNING_COURSE_BUCKET_NAME=mitx-etl-mitxonline-production
```

On master branch:
- On an existing learning resource run (from any source) with at least one existing instructor, add a new instructor with first name "Esther", last name "Duflo", full  name "Esther Duflo".  Save the change.
- On another learning resource run, do the same, with the same full name, but a blank first and last name.  Save the change.
- Run `python manage.py backpopulate_mitxonline_data` - it should fail because of dupe instructors.

On this branch:
- Run `manage.py migrate`
- It should complete without errors.  
- Search `CourseInstructor` objects for  `full_name="Esther Duflo"`.   There should be only one, with a populated first and last name.
- Check that the two runs you modified previously still include "Esther Duflo" as an instructor, along with all the other instructors it had previously.
- Run `python manage.py backpopulate_mitxonline_data`.  It should complete successfully.
